presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java

```diff
@@ -337,40 +348,55 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
+        if (allowMissingFilesForEmptyBuckets) {
+            fileInfos.stream()
+                    .forEach(fileInfo -> bucketToFileInfo.put(getBucketNumber(fileInfo.getPath().getName()), fileInfo));
+        }
+        else {
+            // verify we found one file per bucket
+            if (fileInfos.size() != partitionBucketCount) {
                 throw new PrestoException(
-                        HIVE_INVALID_FILE_NAMES,
-                        format("Hive table '%s' is corrupt. Some of the filenames in the partition: %s are not integers",
+                        HIVE_INVALID_BUCKET_FILES,
+                        format("Hive table '%s' is corrupt. The number of files in the directory (%s) does not match the declared bucket count (%s) for partition: %s",
                                 new SchemaTableName(table.getDatabaseName(), table.getTableName()),
+                                fileInfos.size(),
+                                partitionBucketCount,
                                 partitionName));
             }
-        }
-        else {
-            // Sort FileStatus objects (instead of, e.g., fileStatus.getPath().toString). This matches org.apache.hadoop.hive.ql.metadata.Table.getSortedPaths
-            fileInfos.sort(null);
+
+            if (fileInfos.get(0).getPath().getName().matches("\\d+")) {
+                try {
+                    // File names are integer if they are created when file_renaming_enabled is set to true
+                    fileInfos.sort(Comparator.comparingInt(fileInfo -> Integer.parseInt(fileInfo.getPath().getName())));
+                }
+                catch (NumberFormatException e) {
+                    throw new PrestoException(
+                            HIVE_INVALID_FILE_NAMES,
+                            format("Hive table '%s' is corrupt. Some of the filenames in the partition: %s are not integers",
+                                    new SchemaTableName(table.getDatabaseName(), table.getTableName()),
+                                    partitionName));
+                }
+            }
+            else {
+                // Sort FileStatus objects (instead of, e.g., fileStatus.getPath().toString). This matches org.apache.hadoop.hive.ql.metadata.Table.getSortedPaths
+                fileInfos.sort(null);
+            }
+            for (int i = 0; i < fileInfos.size(); i++) {
+                bucketToFileInfo.put(i, fileInfos.get(i));
+            }
         }
 
         // convert files internal splits
         List<InternalHiveSplit> splitList = new ArrayList<>();
         for (int bucketNumber = 0; bucketNumber < max(readBucketCount, partitionBucketCount); bucketNumber++) {
             // Physical bucket #. This determine file name. It also determines the order of splits in the result.
             int partitionBucketNumber = bucketNumber % partitionBucketCount;
+            if (!bucketToFileInfo.containsKey(partitionBucketNumber)) {
+                continue;
```

Comment:
@wenleix I'm trying to analyze if not producing a split for a bucket wouldn't create a correctness issue. I remember there was some very tricky scenario when we require empty splits to be scheduled for all the buckets to make an aggregation to produce default output. Do you remember exactly what was the scenario and maybe give some suggestion how a test case can be implemented to make sure correctness is not impacted?

Reply:
@arhimondr :  Grouped execution can correctly handle empty buckets (e.g bucket without any splits). A special "EMTPY_SPLIT" will be inserted for each bucket to make sure everything works as expected. 

My understanding is this case is special since even no `InternalHiveSplit` will be generated.  While I don't see any problems there,  I do agree it's better to get this case tested as well. 

I would suggest to have an integration test using exchange materialization, and there is skew so some intermediate partition is empty.

Reply:
@wenleix  @arhimondr  Unit test "testBucketedTemporaryTableWithMissingFiles" is added for this case: the temporary table (11 buckets) has only one row, therefore, 10 buckets are empty without any zero-row file and 1 bucket is non-empty.